const Activity = require("../models/activity");
const Trip = require("../models/trip");

// Creazione di un'attivit√†
exports.createActivity = async (req, res) => {
    try {
        const { itinerario_id, nome, descrizione, luogo, data } = req.body;

        console.log("body", req.body);
      
        // chiedere ad ale se l'ID del viaggio √® fornito (se l'attivit√† √® associata a un viaggio)
        if (itinerario_id) {
            const trip = await Trip.findByPk(itinerario_id);
            if (!trip) {
                return res.status(404).json({ error: 'Viaggio non trovato.' });
            }
        }
        const newActivity = await Activity.create({
            itinerario_id: itinerario_id, // Sar√† nullo se l'attivit√† non √® associata a un viaggio specifico
            nome: nome,
            descrizione: descrizione,
            data: data,
            luogo: luogo
        });
      
        res.status(201).json(newActivity);

    } catch (error) {
        console.error("Errore durante la creazione dell'attivit√†:", error);
        res.status(500).json({ error: "Errore del server durante la creazione dell'attivit√†." });
    }
};

// Recupero delle attivit√† associate a un particolare viaggio
exports.getActivitiesByTrip = async (req, res) => {
    const { tripID } = req.params;

    console.log("üî¥tripID:", tripID);
    try {
      const trip = await Trip.findByPk(tripID);
      if (!trip) {
        return res.status(404).json({ error: 'Viaggio non trovato.' });
      }
      const activities = await Activity.findAll({
        where: { itinerario_id: tripID },
      });

      console.log("üü¢activityController:activities:",activities);
      res.status(200).json(activities);
    } catch (error) {
      console.error(`Errore durante il recupero delle attivit√† per il viaggio con ID ${tripId}:`, error);
      res.status(500).json({ error: 'Errore del server durante il recupero delle attivit√† del viaggio.' });
    }
};

// Recupero di un'attivit√† specifica per id (per eventuali barre di ricerca)
exports.getActivityById = async (req, res) => {
    const { activityID } = req.params;
    try {

        const activity = await Activity.findByPk(activityID);

        if (!activity) return res.status(404).json({ error: 'Attivit√† non trovata.' });

        res.status(200).json(activity);

    } catch (error) {

        console.error(`Errore durante il recupero dell'attivit√† con ID ${activityID}:`, error);
        res.status(500).json({ error: "Errore del server durante il recupero dell'attivit√†." });
    }
};

// Aggiornamento di un'attivit√†
exports.updateActivity = async (req, res) => {
    const { id } = req.params;
    const { name, description, location, date } = req.body;
  
    try {
      const activity = await Activity.findByPk(id);
      if (!activity) {
        return res.status(404).json({ error: 'Attivit√† non trovata.' });
      }

      activity.nome = name;
      activity.descrizione = description;
      activity.data = date;
      activity.luogo = location;
  
      await activity.save();
      res.status(200).json(activity);
    } catch (error) {
        console.error(`Errore durante l'aggiornamento dell'attivit√† con ID ${id}:`, error);
        res.status(500).json({ error: "Errore del server durante l'aggiornamento dell'attivit√†." });
    }
};

// Eliminazione di un'attivit√†
exports.deleteActivity = async (req, res) => {
    const { id } = req.params;
    try {
        const activity = await Activity.findByPk(id);
        if (!activity) {
          return res.status(404).json({ error: 'Attivit√† non trovata.' });
        }
        await activity.destroy();
        res.status(204).send(); // 204 No Content: la richiesta √® andata a buon fine ma non ha contenuto da inviare indietro
    } catch (error) {
        console.error(`Errore durante l'eliminazione dell'attivit√† con ID ${id}:`, error);
        res.status(500).json({ error: "Errore del server durante l'eliminazione dell'attivit√†." });
    }
};
